### ISSUE TITLE
View reference types are ambiguous with strings

### ISSUE THREAD

**Issue by carson-katri** [liveview-native/live_view_native_stylesheet#28]
Some modifiers have arguments that accept either a string, or a reference to an element (for example, `navigationTitle`). A view reference can currently be an :atom, "string" or ["array", "of", "strings"] . However, this is leading to some ambiguity between string arguments and references.

Currently, the SwiftUI client works around this by using the reference as a string if a matching element cannot be found.

**Comment by NduatiK** [comment]
@carson-katri, from the conversation on Slack a while back, it sounds like the issue is that we can't differentiate between strings and atoms in class names.

In the case below, it is unclear whether test is an atom or a string.

```
"navigation-title:" <> title do
  navigationTitle(title)
end
<VStack class="navigation-title:test">
  ...
</VStack>
```

**Comment by NduatiK** [comment]
Assuming the above is true, what are your thoughts on the following:

### Alternative 1

1. We add a modifier for atom types:
```elixir
"navigation-title:" <> title::atom do
  navigationTitle(title) # title is converted to to_atom(title) and sent as an atom allowing you to treat it as a reference
end
```

2. We treat all other cases as strings
```elixir
"navigation-title:" <> title do
  navigationTitle(title) # title is sent as a string
end
```

### Alternative 2

We check for a leading colon:
```elixir
"navigation-title:" <> title do
  navigationTitle(title)
end

<VStack class="navigation-title:test"> # title will have the value "test" and be treated as a string
 ...
</VStack>

<VStack class="navigation-title::test"> # title will have the value ":test" and be converted to an atom
  ...
</VStack>

<VStack class="navigation-title:\:test">
   # title will have the value "\:test" and because the leading colon is escaped,
   # it is treated as the string ":test"
  ...
</VStack>

<VStack class="navigation-title:[attr-title]" title="test">
   # [attr-title] is converted to attr(title)
  ...
</VStack>

```

---

Alternative 2 lets us support the dual behavior expected by views like `navigationTitle`, we can support references or strings.

**Comment by carson-katri** [comment]
I think the issue is more related to allowing strings to be view references. The alternatives only work if we don’t allow string arguments to be interpreted as a view. I’m not sure shorthand should be applied to use defined class names.

Some other possible solutions:
1. Only allow view references to be atoms
2. Add a custom type for view references, like event or attr

**Comment by bcardarella** [comment]
I'd like to avoid implicit type conversions. One of the primary values in Elixir is staying *explicit*. When we start to make implicit type changes while it may serve our immediate needs I've always found it leads to longer-term bad software architecture.

I think the example cited here would be best served using the `attr` function:

```html
<VStack class="navigation-title" title="fooBar">
  ...
</VStack>
```

```elixir
"navigation-title" do
   navigationTitle(attr(title))
end
```

**Comment by bcardarella** [comment]
@carson-katri can you provide a specific example of this in the stylesheet code with a corresponding element?

**Comment by carson-katri** [comment]
```ex
"navigation-title:" <> title do
  navigationTitle(title)
end
```

`title` is passed as a string. The Swift side tries to interpret this based on these two signatures:

```
navigationTitle(_ title: ViewReference)
navigationTitle(_ title: String)
```

`ViewReference` can be created from a string, atom, or list of strings/atoms. So it resolves to this. However, the users intent in this case was to use the second signature that just uses the string directly. However, it is unreachable because of the ordering and allowed forms of `ViewReference`.

**Comment by carson-katri** [comment]
Yes, that's one possible solution. The downside is it makes cases where there is no ambiguity more verbose. For example, `background` doesn't have this issue, but I would still have to write out

```ex
"bg:" <> content do
  background(content: to_atom(content))
end
```

**Comment by bcardarella** [comment]
as I stated previously in this thread I think we should stick to being explicit. Which means we should always explicitly type coerce to an `Atom` for any `ViewReference`.
